In [1]:
import math
import numpy as np

https://converged.yt/primel/ 

In [2]:
max_num = 99999
 
def get_primes():
    primes = []
    for num in range(3,max_num,2):
        if all(num%i!=0 for i in range(3,int(math.sqrt(num))+1, 2)):
            primes.append(num)
    return primes

In [3]:
primes = get_primes()

filtered = list(filter(lambda primes: primes >= 9999, primes)) # filter primes with <5 digits

primes = filtered

In [4]:
## best primes for initial guess (5 different digits)
start_primes = []
for prime in primes:
    digit_list = [int(x) for x in str(prime)] # creat list of int
    #print(len(list(dict.fromkeys(digit_list))))
    if len(list(dict.fromkeys(digit_list))) == 5: # only for 5 different digits append
        start_primes.append(prime)
print(start_primes[-5:])  

        

[98621, 98627, 98641, 98713, 98731]


In [5]:
digit_count = np.zeros(10)

for prime in primes:
    digit_list = [int(x) for x in str(prime)] # creat list of int
    for digit in digit_list:
        digit_count[digit] = digit_count[digit] + 1
for i, digit in enumerate(digit_count):       
    print(i, ": ", digit_count[i] / len(primes))
    
# best digits to include for start guess: 1, 3, 7, 9

incl_numbers = [1, 3, 7, 9, 2]

temp_primes = primes
for incl_number in incl_numbers:
    best_start_primes = []
    for prime in temp_primes:
        for digit in str(prime):
            if int(digit) == incl_number:
                best_start_primes.append(prime)
    temp_primes = best_start_primes

best_start_primes = temp_primes
best_start_primes = list(dict.fromkeys(best_start_primes)) 

print(best_start_primes)

0 :  0.298098768384551
1 :  0.6782255171589142
2 :  0.4203037187612101
3 :  0.6638765993064689
4 :  0.40798756427119454
5 :  0.4132488341504245
6 :  0.4032045916537128
7 :  0.6600502212124836
8 :  0.39925863924429034
9 :  0.6557455458567499
[12379, 12739, 12973, 13297, 13729, 17239, 17293, 17923, 19237, 19273, 21379, 21397, 21739, 21937, 23197, 23719, 23917, 23971, 29137, 29173, 31729, 32719, 32917, 32971, 39217, 71293, 71329, 72139, 72931, 73291, 79231, 91237, 92173, 92317, 97213, 97231]


In [6]:
# TRY:       71293  
# YELLOW:    7, 1, 9
# GREY:      2, 3


results = [np.NaN, np.NaN, np.NaN, np.NaN, np.NaN] # insert np.NaN for not correct digits, and correct digits in the list
incl_numbers = [7, 1, 9] # put yellow marked digits, and digits you havent tried yet
excl_numbers = [2, 3] # put grey marked digits in here

temp_primes = []

for i, result in enumerate(results):
    if result is not np.NaN:
        if not temp_primes:
            for prime in primes:
                if int(str(prime)[i]) is result:
                    temp_primes.append(prime)
        else:
            new_primes = []
            for prime in temp_primes:
                if int(str(prime)[i]) is result:
                    new_primes.append(prime)
            temp_primes = new_primes

if all(result is np.NaN for result in results): temp_primes = primes            

for incl_number in incl_numbers:
    new_primes = []
    for prime in temp_primes:
        for digit in str(prime):
            if int(digit) == incl_number:
                new_primes.append(prime)
        temp_primes = new_primes
        
for excl_number in excl_numbers:
    new_primes = []
    for prime in temp_primes:
        append = True
        for digit in str(prime):
            if int(digit) == excl_number:
                append = False 
        if append:    
            new_primes.append(prime)
        temp_primes = new_primes
        
temp_primes = list(dict.fromkeys(temp_primes))            
print(temp_primes) # prints possilbe guesses, use one which includes digits you havent tried yet

[10079, 10597, 10709, 10789, 10799, 10957, 10979, 10987, 11197, 11497, 11579, 11597, 11719, 11779, 11789, 11897, 11971, 11987, 14197, 14479, 14759, 14779, 14797, 14879, 14897, 14947, 14957, 15497, 15679, 15749, 15791, 15797, 15907, 15971, 16097, 16759, 16879, 16979, 16987, 17099, 17159, 17189, 17191, 17419, 17449, 17489, 17491, 17497, 17509, 17519, 17569, 17579, 17597, 17599, 17609, 17659, 17669, 17749, 17789, 17791, 17891, 17909, 17911, 17957, 17959, 17971, 17977, 17981, 17987, 17989, 18097, 18679, 18719, 18749, 18797, 18917, 18947, 18979, 19079, 19087, 19157, 19417, 19447, 19457, 19471, 19477, 19507, 19571, 19577, 19597, 19687, 19697, 19709, 19717, 19751, 19759, 19777, 19867, 19979, 19997, 41179, 41479, 41579, 41597, 41719, 41759, 41879, 41897, 41947, 41957, 44179, 44917, 44971, 45179, 45197, 45971, 47119, 47149, 47189, 47419, 47491, 47591, 47791, 47819, 47911, 47917, 47951, 47981, 48179, 48197, 49117, 49157, 49171, 49177, 49417, 49711, 49741, 49871, 50971, 51197, 51479, 51679, 51719

In [7]:
# TRY:       59671  
# YELLOW:    7, 1, 9
# GREY:      2, 3, 5, 6


results = [np.NaN, np.NaN, np.NaN, 7, np.NaN] # insert np.NaN for not correct digits, and correct digits in the list
incl_numbers = [7, 1, 9] # put yellow marked digits, and digits you havent tried yet
excl_numbers = [2, 3, 5, 6] # put grey marked digits in here

temp_primes = []

for i, result in enumerate(results):
    if result is not np.NaN:
        if not temp_primes:
            for prime in primes:
                if int(str(prime)[i]) is result:
                    temp_primes.append(prime)
        else:
            new_primes = []
            for prime in temp_primes:
                if int(str(prime)[i]) is result:
                    new_primes.append(prime)
            temp_primes = new_primes

if all(result is np.NaN for result in results): temp_primes = primes            

for incl_number in incl_numbers:
    new_primes = []
    for prime in temp_primes:
        for digit in str(prime):
            if int(digit) == incl_number:
                new_primes.append(prime)
        temp_primes = new_primes
        
for excl_number in excl_numbers:
    new_primes = []
    for prime in temp_primes:
        append = True
        for digit in str(prime):
            if int(digit) == excl_number:
                append = False 
        if append:    
            new_primes.append(prime)
        temp_primes = new_primes
        
temp_primes = list(dict.fromkeys(temp_primes))            
print(temp_primes) # prints possilbe guesses, use one which includes digits you havent tried yet

[10079, 10979, 11779, 11971, 14479, 14779, 14879, 17971, 17977, 18979, 19079, 19471, 19477, 19777, 19979, 41179, 41479, 41879, 44179, 44971, 48179, 49171, 49177, 49871, 71479, 71879, 71971, 78179, 81971, 84179, 87179, 89071, 90071, 90971, 91079, 91771, 94771, 97171, 97177, 97771, 97871, 98179, 99871, 99971]


In [8]:
# TRY:       48179  
# YELLOW:    7, 1, 9, 4
# GREY:      2, 3, 5, 6, 8


results = [np.NaN, np.NaN, np.NaN, 7, 9] # insert np.NaN for not correct digits, and correct digits in the list
incl_numbers = [7, 1, 9, 4] # put yellow marked digits, and digits you havent tried yet
excl_numbers = [2, 3, 5, 6, 8] # put grey marked digits in here

temp_primes = []

for i, result in enumerate(results):
    if result is not np.NaN:
        if not temp_primes:
            for prime in primes:
                if int(str(prime)[i]) is result:
                    temp_primes.append(prime)
        else:
            new_primes = []
            for prime in temp_primes:
                if int(str(prime)[i]) is result:
                    new_primes.append(prime)
            temp_primes = new_primes

if all(result is np.NaN for result in results): temp_primes = primes            

for incl_number in incl_numbers:
    new_primes = []
    for prime in temp_primes:
        for digit in str(prime):
            if int(digit) == incl_number:
                new_primes.append(prime)
        temp_primes = new_primes
        
for excl_number in excl_numbers:
    new_primes = []
    for prime in temp_primes:
        append = True
        for digit in str(prime):
            if int(digit) == excl_number:
                append = False 
        if append:    
            new_primes.append(prime)
        temp_primes = new_primes
        
temp_primes = list(dict.fromkeys(temp_primes))            
print(temp_primes) # prints possilbe guesses, use one which includes digits you havent tried yet

[14479, 14779, 41179, 41479, 44179, 71479]


[14479, 14779, 41179, 41479, 44179, 71479]
     p,     p,    np,    np,    np,    np 

With exclusion principle: 14479 or 14779

14779 correct!

### ToDo
* Get most frequent digits for initial guess --> Done
* Don't use already tried digits --> Done (grey digits)
* Don't repeat already used position for yellow